### ISSUE TITLE
method.AUC gives bad coefs when run twice

### ISSUE THREAD

**Issue by ck37** [ecpolley/SuperLearner#30]
Hi,

I'm running into a very strange issue where method.AUC works the first time it's used in a SuperLearner, but the second or later times it returns coefficients that are outside [0, 1] and result in invalid predictions. This is being run on exactly the same data each time.

Here is code that reproduces the issue:

```{r}
library(SuperLearner)
library(glmnet)
library(cvAUC)

set.seed(1)
N <- 200
X <- matrix(rnorm(N*10), N, 10)
X <- as.data.frame(X)
Y <- rbinom(N, 1, plogis(.2*X[, 1] + .1*X[, 2] - .2*X[, 3] + .1*X[, 3]*X[, 4] - .2*abs(X[, 4])))
table(Y)

SL.library <- c("SL.glmnet", "SL.mean")

test.AUC <- SuperLearner(Y = Y, X = X, SL.library = SL.library, verbose = F, method = "method.AUC", family = binomial())
test.AUC
summary(test.AUC$SL.predict)

# SL coefficients are not bounded by [0, 1] and result in predictions outside bounds of Y.
test.AUC2 <- SuperLearner(Y = Y, X = X, SL.library = SL.library, verbose = F, method = "method.AUC", family = binomial())
test.AUC2
summary(test.AUC2$SL.predict)

# Third try doesn't work either :(
test.AUC3 <- SuperLearner(Y = Y, X = X, SL.library = SL.library, verbose = F, method = "method.AUC", family = binomial())
test.AUC3
summary(test.AUC3$SL.predict)



# These NNLS versions are fine.
test.NNLS <- SuperLearner(Y = Y, X = X, SL.library = SL.library, verbose = F, method = "method.NNLS", family = binomial())
test.NNLS
summary(test.NNLS$SL.predict)

test.NNLS2 <- SuperLearner(Y = Y, X = X, SL.library = SL.library, verbose = F, method = "method.NNLS", family = binomial())
test.NNLS2
summary(test.NNLS2$SL.predict)

```

This is using R 3.2.5 and SuperLearner 2.0-20 (package created on 2016-04-06).

Thanks,
Chris

**Comment by lendle** [comment]
In your example, the change across runs is due to the randomness in cross validation splits, and coefficients only happen to be in [0,1] by chance. 

Code here:
https://github.com/ecpolley/SuperLearner/blob/master/R/method.R#L254-L262

Predictions when using method.AUC are a linear combination of the predictions from the candidates. The coefficients are not constrained to be non-negative or sum to 1.

**Comment by ecpolley** [comment]
I agree with what Sam said above, and will add that you could specify the cross-validation splits in the `cvControl` argument so that would be consistent across runs, and you would also need to do the same in `SL.glmnet` (see the `foldid` argument in `cv.glmnet`).

The other way is to specify the seed just prior to each of the `SuperLearner` calls.

    set.seed(1)
    A <- SuperLearner(Y = Y, X = X, SL.library = SL.library, verbose = F, method = "method.AUC", family = binomial())

    set.seed(1)
    B <- SuperLearner(Y = Y, X = X, SL.library = SL.library, verbose = F, method = "method.AUC", family = binomial())

    identical(coef(A), coef(B))
    identical(A$cvRisk,B$cvRisk)

**Comment by ck37** [comment]
Ah, that explains it - apologies for not realizing that earlier.

My optimization is rusty, but I don't follow why the resulting weights are useful if they aren't constrained to [0, 1].

Here is a simple example, tweaked from above to add a set.seed(1) before SL:
```{r}
library(SuperLearner)
library(glmnet)
library(cvAUC)

rm(list=ls())

set.seed(1)
N <- 200
X <- matrix(rnorm(N*10), N, 10)
X <- as.data.frame(X)
Y <- rbinom(N, 1, plogis(.2*X[, 1] + .1*X[, 2] - .2*X[, 3] + .1*X[, 3]*X[, 4] - .2*abs(X[, 4])))
table(Y)

SL.library <- c("SL.glmnet", "SL.mean")

set.seed(1)
test.AUC <- SuperLearner(Y = Y, X = X, SL.library = SL.library, verbose = T, method = "method.AUC", family = binomial())
test.AUC
summary(test.AUC$SL.predict)
```

The distribution of the predicted values is:
```{r}
> summary(test.AUC$SL.predict)
       V1         
 Min.   :-5.6535  
 1st Qu.:-3.4774  
 Median :-2.7330  
 Mean   :-2.7291  
 3rd Qu.:-2.0309  
 Max.   : 0.5566 
```
Only 1% of the predictions are between 0 and 1. 99% are invalid predictions and method.AUC is only intended to work for the binomial family.

Unlike the other methods, method.AUC does not output the convergence status of the optim() call, but these bad coefficients and predictions appear to occur when the nelder-mead optimization does not converge. Whereas when it does converge it returns valid predictions and coefficients in [0, 1]. So I agree that is is chance that the algorithm converges and gives valid results in my first attempt, but the convergence of the optimization algorithm can be checked and if it doesn't converge it seems to give bad results. 

I could definitely be missing something here (apologies if so) but I don't understand how this method is useable right now.

**Comment by ck37** [comment]
Switching the optimization from nelder-mead to L-BFGS-B and adding [0, 1] bounds seems to resolve the issue:

```{r}
library(SuperLearner)
library(glmnet)
library(cvAUC)

set.seed(1)
N <- 200
X <- matrix(rnorm(N*10), N, 10)
X <- as.data.frame(X)
Y <- rbinom(N, 1, plogis(.2*X[, 1] + .1*X[, 2] - .2*X[, 3] + .1*X[, 3]*X[, 4] - .2*abs(X[, 4])))
table(Y)

SL.library <- c("SL.glmnet", "SL.mean")

# Works fine by luck at converge.
test.AUC <- SuperLearner(Y = Y, X = X, SL.library = SL.library, verbose = F, method = "method.AUC", family = binomial())
test.AUC
summary(test.AUC$SL.predict)

# Fails due to lack of convergence.
set.seed(1)
test.AUC2 <- SuperLearner(Y = Y, X = X, SL.library = SL.library, verbose = F, method = "method.AUC", family = binomial())
test.AUC2
summary(test.AUC2$SL.predict)

method.AUC2 <- function(optim_method="L-BFGS-B") {
  # Contributed by Erin Ledell
  out <- list(
    require = c('cvAUC', 'ROCR'),
    computeCoef = function(Z, Y, libraryNames, obsWeights, control, verbose, ...) {
      .cvRisk_AUC <- function(par, Z, Y, folds=NULL){
        # Calculate cvRisk, which is 1-cvAUC (Rank Loss)
        # This is the loss function that gets fed into optim as the "fn" argument
        # par is the weight/coef vector for the ensemble in Super Learner
        predictions <- crossprod(t(Z), par)  #cv predicted SL values
        cvRisk <- 1 - cvAUC::cvAUC(predictions=predictions, labels=Y, folds=folds)$cvAUC
        return(cvRisk)
      }
      coef_init <- rep(1/ncol(Z),ncol(Z))
      names(coef_init) <- libraryNames
      # optim function selects the value for par that minimizes cvRisk_AUC (aka Rank Loss)
      #res <- optim(par=coef_init, fn=.cvRisk_AUC, Z=Z, Y=Y, folds=NULL, method=optim_method)
      res <- optim(par=coef_init, fn=.cvRisk_AUC, Z=Z, Y=Y, folds=NULL, method=optim_method, lower=0, upper=1)
      if(verbose) {
        message(paste("AUC convergence: ", res$convergence == 0))
      }
      coef <- res$par
      auc <- apply(Z, 2, function(x) cvAUC::cvAUC(x, labels=Y)$cvAUC)
      cvRisk <- 1 - auc  # Rank Loss
      names(coef) <- libraryNames
      out <- list(cvRisk = cvRisk, coef = coef, optim_result = res)
      return(out)
    },
    computePred = function(predY, coef, control, ...) {
      out <- crossprod(t(predY), coef)
      return(out)
    }
  )
  invisible(out)
}

# Should work.
set.seed(1)
test.AUC3 <- SuperLearner(Y = Y, X = X, SL.library = SL.library, verbose = F, method = "method.AUC2", family = binomial())
test.AUC3
summary(test.AUC3$SL.predict)
```

**Comment by ecpolley** [comment]
There is no theoretical reason the coefficients need to be restricted to between 0 and 1. We initially chose the convex combination (alpha > 0 and sum = 1) in the binary outcome setting because it was helpful for the interpretation, note that in the NNLogLik the ensemble is also on the logit scale, not the probability scale as is used in the AUC code above.

The other action that will likely help the problem is increasing the number of algorithms in SL.library (I know you already stated you case for starting with a small library, but this step in the SuperLearner code isn't really designed for a single algorithm). SL.mean is a constant (intercept term).

**Comment by ecpolley** [comment]
Thanks for noting that optim doesn't give any feedback when it doesn't converge, it is a nice suggestion to add a warning message about this in method.AUC.

**Comment by ecpolley** [comment]
Chris, would it be OK if I added your method.AUC2 as an example in the SuperLearnerExtra folder?

https://github.com/ecpolley/SuperLearnerExtra/tree/master/method

**Comment by lendle** [comment]
Re: why aren't predictions in [0,1]? 

AUC is kind of a weird metric because it only depends on the relative order of predictions, not value. This means that from the optimizer's perspective, any vector of coefficients that result in the same order of predictions are equivalent. Because of this, predictions when using method.AUC can't be interpreted as a probability.

**Comment by ecpolley** [comment]
Hi Sam, the predictions outside [0,1] are the result of the linear model assumed not the loss function. In the NNLogLik (and CC_nloglik) the ensemble is on the logit scale and then transformed back into [0,1] by the inverse logit function. This forces the predictions to be between 0 and 1. In method.AUC, the ensemble model is additive without the transformations so can be outside the bounds.

**Comment by ck37** [comment]
Yes, definitely feel free to add method.AUC2 to anything.

Adding the warning is handy but it would be even more helpful for each SL meta-learner to return the convergence result of the optimizer so that it can be checked programmatically instead of having to parse warning messages. This seems a lot more important if the meta-learner has a reasonable probability of not converging.

Adding more learners makes sense, but is that helpful because the optimizer would be more likely to move to a convex weighting than away from it?

**Comment by ledell** [comment]
Just noticed this thread... @ck37 I have some updates to `method.AUC` that should improve performance.  I did not have much luck with Nelder-Mead (this should not be the default).  The only method provided by `optim()` that did an okay job was L-BFGS-B, but the performance was still not great.  In response, I expanded `method.AUC` to utilize a large number of constrained optimization methods from the [https://github.com/jyypma/nloptr](nloptr) library.

I haven't updated the code on GitHub yet, but here is a [the updated method](https://gist.github.com/ledell/66bab5820b7fb5a0d4f239f29069e937) you can try out, and here is some [code](https://gist.github.com/ledell/1e79ae4042e73146dbfd6eaaf8c0079e) for generating a bunch of different versions of the method that use different optimization methods.

**Comment by ck37** [comment]
Thanks @ledell this is super helpful, much appreciated.

**Comment by ledell** [comment]
@ck37 There are some details in [Chapter 4](http://www.stat.berkeley.edu/~ledell/papers/ledell-phd-thesis.pdf) of my dissertation. (Which shows how long I have been sitting on this code...oops!!)

**Comment by ck37** [comment]
Whoa awesome chapter, shoulda found that one on my own! Looks like my hunch optimizer, L-BFGS-B with [0, 1] bounds, was not too shabby as AUC_optim.1.

**Comment by ecpolley** [comment]
Thanks everyone! I've updated the method.AUC with the new defaults (L-BFGS-B and [0, inf) bounds) and have included the warning messages. I agree it would be nice to have convergence stored programmatically, and have added this to the ToDo list, but this will require some edits to the SuperLearner function. The solution suggested above about adding to the output from method$computeCoef() is a good start, but this information doesn't directly get returned from the SuperLearner() call and will be lost.

**Comment by ck37** [comment]
Yep, I modified my local SuperLearner source to return the result while debugging. I'll take a stab at a pull request for a real patch.
